jetstream.go

```diff
@@ -0,0 +1,334 @@
+		return ErrMsgNoReply
+	}
+
+	if m == nil || m.Sub == nil {
+		return ErrMsgNotBound
+	}
+
+	m.Sub.mu.Lock()
+	nc := m.Sub.conn
+	m.Sub.mu.Unlock()
+
+	var err error
+
+	aopts := newJsOpts()
+	for _, f := range opts {
+		err = f(aopts)
+		if err != nil {
+			return err
+		}
+	}
+
+	if aopts.timeout == 0 {
+		return m.Respond(body)
+	}
+
+	_, err = nc.Request(m.Reply, body, aopts.timeout)
+
+	return err
+}
+
+// Ack acknowledges a JetStream messages received from a Consumer, indicating the message
+// should not be received again later
+func (m *Msg) Ack(opts ...AckOption) error {
+	return m.jsAck(AckAck, opts...)
+}
+
+// Nak acknowledges a JetStream message received from a Consumer, indicating that the message
+// is not completely processed and should be sent again later
+func (m *Msg) Nak(opts ...AckOption) error {
+	return m.jsAck(AckNak, opts...)
+}
+
+// AckProgress acknowledges a Jetstream message received from a Consumer, indicating that work is
+// ongoing and further processing time is required equal to the configured AckWait of the Consumer
+func (m *Msg) AckProgress(opts ...AckOption) error {
+	return m.jsAck(AckProgress, opts...)
+}
+
+// AckNext performs an Ack() and request that the next message be sent to subject ib
+func (m *Msg) AckNext(ib string) error {
+	return m.RespondMsg(&Msg{Subject: m.Reply, Reply: ib, Data: AckNext})
+}
+
+// AckTerm acknowledges a message received from JetStream indicating the message will not be processed
+// and should not be sent to another consumer
+func (m *Msg) AckTerm(opts ...AckOption) error {
+	return m.jsAck(AckTerm, opts...)
+}
+
+// JetStreamPublishAck metadata received from JetStream when publishing messages
+type JetStreamPublishAck struct {
+	Stream   string `json:"stream"`
+	Sequence int    `json:"seq"`
+}
+
+// ParsePublishAck parses the publish acknowledgement sent by JetStream
+func ParsePublishAck(m []byte) (*JetStreamPublishAck, error) {
+	if bytes.HasPrefix([]byte("-ERR"), m) {
+		if len(m) > 7 {
+			return nil, fmt.Errorf(string(m[6 : len(m)-1]))
+		}
+
+		return nil, fmt.Errorf(string(m))
+	}
+
+	if !bytes.HasPrefix(m, []byte("+OK {")) {
+		return nil, fmt.Errorf("invalid JetStream Ack: %v", string(m))
+	}
+
+	ack := &JetStreamPublishAck{}
+	err := json.Unmarshal(m[3:], ack)
+	return ack, err
+}
+
+func isValidJSName(n string) bool {
+	if n == "" {
+		return false
+	}
+
+	return !strings.ContainsAny(n, ">*. ")
+}
+
+func (nc *Conn) jsPublish(subj string, data []byte, opts []PublishOption) error {
+	aopts := newJsOpts()
+	var err error
+
+	for _, f := range opts {
+		err = f(aopts)
+		if err != nil {
+			return err
+		}
+	}
+
+	if aopts.timeout == 0 && aopts.ctx == nil && aopts.ack.str == jsStreamUnspecified {
+		return nc.publish(subj, _EMPTY_, nil, data)
+	}
+
+	ctx, cancel := aopts.context(nc.Opts.JetStreamTimeout)
+	defer cancel()
+
+	resp, err := nc.RequestWithContext(ctx, subj, data)
+	if err != nil {
+		return err
+	}
+
+	ack, err := ParsePublishAck(resp.Data)
```

Comment:
We should only do this if the pubAck option included a stream name, otherwise just check prefix for +OK or -ERR IMO.

Reply:
The first thing ParsePublishAck does is checks for -ERR and if its there pulls out the error message and returns that as `err` here, this way when someone does nc.Publish() the err they get back holds the jetstream supplied error message.

Reply:
Right but if I just care that it said OK I am parsing and creating alot of stuff for nothing.

Reply:
OK , will do. Curious though what we're optimising this so heavily for, this is after a `Request()` that did network round trips, flush  etc, seems like a distraction, this isnt high performance code paths

Reply:
Was suggestion, you don't have to put it in.